lib/common/common/src/mmap_hashmap.rs

```diff
@@ -0,0 +1,233 @@
+use std::collections::{BTreeMap, BTreeSet};
+use std::fs::File;
+use std::io::{self, Cursor, Write as _};
+use std::mem::size_of;
+use std::path::Path;
+use std::str;
+
+use memmap2::Mmap;
+use ph::fmph::Function;
+#[cfg(test)]
+use rand::rngs::StdRng;
+#[cfg(test)]
+use rand::{Rng as _, SeedableRng as _};
+use zerocopy::{AsBytes, FromBytes, FromZeroes};
+
+/// On-disk hash map baked by a memory-mapped file.
+pub struct MmapHashMap {
+    mmap: Mmap,
+    header: Header,
+    phf: Function,
+}
+
+#[repr(C)]
+#[derive(Copy, Clone, Debug, AsBytes, FromBytes, FromZeroes)]
+struct Header {
+    buckets_pos: u64,
+    buckets_count: u64,
+}
+
+const PADDING_SIZE: usize = 4096;
+
+impl MmapHashMap {
+    /// Save `map` contents to `path`.
+    pub fn create(path: &Path, map: BTreeMap<String, BTreeSet<u32>>) -> io::Result<()> {
+        let phf = Function::from(map.keys().collect::<Vec<_>>().as_slice());
+
+        // == First pass ==
+
+        let mut file_size = 0;
+        // 1. Header
+        file_size += size_of::<Header>();
+
+        // 2. PHF
+        file_size += phf.write_bytes();
+
+        // 3. Padding
+        let padding_len = (PADDING_SIZE - (file_size % PADDING_SIZE)) % PADDING_SIZE;
+        file_size += padding_len;
+
+        // 4. Buckets
+        let buckets_pos = file_size;
+        file_size += map.len() * size_of::<u64>();
+
+        // 5. Data
+        let mut buckets = vec![0u64; map.len()];
+        let mut last_bucket = 0;
+        for (k, v) in map.iter() {
+            buckets[phf.get(k).expect("Key not found in phf") as usize] = last_bucket as u64;
+            last_bucket += entry_size(k, v);
+        }
+        file_size += last_bucket;
+        _ = file_size;
+
+        // == Second pass ==
+
+        let mut bufw = io::BufWriter::new(File::create(path)?);
```

Comment:
We calculated size already. I propose to create file using `create_and_ensure_length` which creates a file with provided size